specs/plugin-extensibility.md

```diff
@@ -155,53 +171,122 @@ This interface targets plugins that integrate with providers of basic cryptograp
     3. Return an error
 
-***generate-signature***
+#### describe-key
+
+This command is used to get metadata for a given key.
 
 *Request*
 
 ```jsonc
 {
   "contractVersion" : "<major-version.minor-version>",
 
-  // Complete key definition from
-  // /notation/config.json /signingKeys/keys with matching key name
-  "keyDefinition" : "<key definition>",
+  // Key name, key id for the key associated with the plugin
+  // in config.json /signingKeys/keys
+  "keyName": "<key name>",
+  "keyId": "<key id>",
+
+  // Optional plugin configuration, map of string-string
+  "pluginConfig" : { }
+}
+```
+
+*keyName* and *keyId* : Required fields that contain the friendly signing key name (`keyName`) and key identifier (`keyId`) as specified in `config.json`.
 
-  "payload" : "<Base64 encoded payload to be signed>"
+*pluginConfig* : Optional field for plugin configuration. For details, see [Plugin Configuration](#plugin-configuration) section.
+
+*Response*
+
+```jsonc
+{
+   // The same key id as passed in the request.
+  "keyId" : "<key id>",
+  "keySpec" : "<key type and size>"
 }
 ```
 
-*keyDefinition* : Required field that contains the complete key definition, which includes friendly name (`name`), key identifier (`id`) and any additional custom fields required by the plugin.
-For example, after registering a key `mysigningkey` and plugin, the command `notation sign $IMAGE --key "mysigningkey"` will subsequently invoke plugin with `request.keyDefinition` set to following value.
+*keySpec* : One of following [supported key types](https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#algorithm-selection) - `RSA_2048`, `RSA_3072`, `RSA_4096`, `EC_256`, `EC_384`, `EC_512`.
+
+NOTE: This command can also be used as part of `notation key describe {key-name}` which will include the following output
+
+* Summary of key definition from `config.json`
+* If the key has an associated plugin
+  * Output of plugin `discover` command
+  * Output of `describe-key` command for the specific key
+
+#### generate-signature
+
+This command is used to generate the raw signature for a given payload.
+
+*Request*
 
 ```jsonc
-{  
-  "name": "mysigningkey",  
-  "id" : "keyid",
-  "plugin": "com.example.nv2plugin"
+{
+  "contractVersion" : "<major-version.minor-version>",
+
+  // Key name, key id for the key associated with the plugin
+  // in config.json /signingKeys/keys
+  "keyName": "<key name>",
+  "keyId": "<key id>",
+
+
+  // Optional plugin configuration, map of string-string
+  "pluginConfig" : { },
+
+  // The key spec for the given key id
+  "keySpec" : "<key type and size>",
+
+  // Hash algorithm associated with the key spec, plugin must 
+  // hash the payload using this hash algorithm
+  "hashAlgorithm" : "SHA_256" | "SHA_384" | "SHA_512",
+
+  // The signature envelope type associated with the payload.
+  "signatureEnvelopeType" : "application/vnd.cncf.notary.v2.jws.v1",
```

Comment:
This used to be only necessary for the generate envelope request. Why would a plugin which is not envelope-aware want to know the envelope type?

Reply:
Generally this would not be required. Currently the encoding of the `payload` field depends on envelope type being targeted. JWS payload is already base64 encoded and signature is generated over a base64 encoded payload, whereas a binary payload would have to be encoded when sending request and decoded by plugin before it generates the raw signature. This is a shortcoming of using JSON as the request encoding, when we switch to something binary friendly like COSE, `signatureEnvelopeType` will not be required in this request.

Reply:
Wouldn't it be simpler if `payload` is always base64-encoded? Making the encoding depend on the envelope type, means that plugins implementing the `generate-signature` workflows will have to be updated if we add a new envelope type, instead of being generic enough to blindly sign whatever payload.

The only advantage I see on having the encoding depending on the envelope format is to avoid double-encoding payloads, for example in JWS. On the other hand, base64 encoding is normally highly optimized, my computer can encode bytes at 500MB/s, so no-one will notice that we are double-encoding some payloads.

Reply:
Agree that including `signatureEnvelopeType` makes the plugin implementing `generate-signature` interface to be updated to support newer payload types, which is not one of the goals of this interface. Consistent payload encoding will in be inline with request contract we would have when we switch to a binary friendly encoding in future.

Reply:
@qmuntal , latest rev includes this change.